### PR TITLE
tag has exported because we need it for writing custom widgets

### PR DESCRIPTION
**Pull request by nervgh**
I'm trying to implement the "file" widget and I need internal tag module with its functionality.

For example:
```js
function widgetFile(options) {
  let opt = options || {}
  let w = {
    classes: opt.classes,
    type: 'file'
  }
  let userAttrs = getUserAttrs(opt) // <-- (1)
  w.toHTML = function (name, field) {
    let f = field || {}
    let attrs = {
      type: 'file',
      name: name,
      id: f.id === false ? false : (f.id || true),
      classes: w.classes
    }
    return tag('input', [attrs, userAttrs, w.attrs || {}], f.value || '') // <-- (2)
  }
  return w
}
```
After my changes I could be write:
```js
let tag = require('forms').tag // <-- (0)

function widgetFile(options) {
  let opt = options || {}
  let w = {
    classes: opt.classes,
    type: 'file'
  }
  let userAttrs = tag.getUserAttrs(opt) // <-- (1)
  w.toHTML = function (name, field) {
    let f = field || {}
    let attrs = {
      type: 'file',
      name: name,
      id: f.id === false ? false : (f.id || true),
      classes: w.classes
    }
    return tag('input', [attrs, userAttrs, w.attrs || {}], f.value || '') // <-- (2)
  }
  return w
}
```

All what doing this PR is exporting "tag" module with its helpers for painless writing custom widgets.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
+.idea/
```
Comment (ljharb) [inline .gitignore]: please revert this; your personal IDE choices should go in your global gitconfig, instead of polluting individual repos

File: lib/forms.js [hunk 2]
```diff
@@ -15,6 +15,7 @@ exports.widgets = require('./widgets');
 exports.fields = require('./fields');
 exports.render = require('./render');
 exports.validators = require('./validators');
+exports.tag = require('./tag');
```
Comment (ljharb) [inline lib/forms.js]: This isn't necessary - `require('forms/lib/tag')` works now. Please revert this.

File: lib/widgets.js [hunk 3]
```diff
@@ -5,24 +5,11 @@ var reduce = require('reduce');
 var some = require('array.prototype.some');
 var tag = require('./tag');
 
-var dataRegExp = /^data-[a-z]+([-][a-z]+)*$/;
-var ariaRegExp = /^aria-[a-z]+$/;
-var legalAttrs = ['autocomplete', 'autocorrect', 'autofocus', 'autosuggest', 'checked', 'dirname', 'disabled', 'tabindex', 'list', 'max', 'maxlength', 'min', 'novalidate', 'pattern', 'placeholder', 'readonly', 'required', 'size', 'step'];
-var ignoreAttrs = ['id', 'name', 'class', 'classes', 'type', 'value', 'multiple'];
-var getUserAttrs = function (opt) {
```
Comment (ljharb) [inline lib/widgets.js]: please do `var getUserAttrs = tag.getUserAttrs;` here to avoid breaking if someone mutates the `tag` object later.

File: .gitignore (hunk 1 above)
Comment (nervgh) [inline .gitignore]: done

File: lib/forms.js (hunk 2 above)
Comment (nervgh) [inline lib/forms.js]: done

File: lib/widgets.js (hunk 3 above)
Comment (nervgh) [inline lib/widgets.js]: done

File: lib/tag.js [hunk 4]
```diff
@@ -60,6 +60,27 @@ var tag = function tag(tagName, attrsMap, content, contentIsEscaped) {
...
+    return reduce(opt, function (attributes, option, k) {
+        if ((ignoreAttrs.indexOf(k) === -1 && legalAttrs.indexOf(k) > -1) || dataRegExp.test(k) || ariaRegExp.test(k)) {
+            attributes[k] = option;
+        }
+        return attributes;
+    }, {});
+};
+
+// expose
+tag.getUserAttrs = getUserAttrs;
```
Comment (ljharb) [inline lib/tag.js]: This looks good, but if it's exposed, it now needs unit tests.

Comment (nervgh) [inline lib/tag.js]: Thank you for your review. I've added a couple of tests :smiley_cat:

Comment (ljharb) [inline lib/tag.js]: Good start! I was hoping for tests that cover all of the functionality.

File: test/test-tag.js [hunk 5]
```diff
@@ -48,3 +48,14 @@ test('allow empty attributes', function (t) {
...
     t.end();
 });
+
+test('should expose getUserAttrs()', function (t) {
+    t.equal(typeof tag.getUserAttrs, 'function');
+    t.end();
+});
+
+test('tag.getUserAttrs() should get and return an object', function (t) {
+    t.throws(tag.getUserAttrs); // eslint-disable-line
```
Comment (ljharb) [inline test/test-tag.js]: why is this line disabled for eslint?

Comment (nervgh) [inline test/test-tag.js]: this line causes a eslint notice

File: lib/tag.js (hunk 4 above)
Comment (nervgh) [inline lib/tag.js]: Sorry, I don't know all use cases so I cannot cover them all.

File: test/test-tag.js (hunk 5 above)
Comment (ljharb) [inline test/test-tag.js]: Which notice? We should try to fix it instead of just overriding.

File: lib/tag.js (hunk 4 above)
Comment (ljharb) [inline lib/tag.js]: If you want to expose a function, then it's worth looking at the code to understand its use cases - especially when that enables you to write tests.

File: test/test-tag.js (hunk 5 above)
Comment (nervgh) [inline test/test-tag.js]: The notice is

> /my_path/caolan.forms/test/test-tag.js
>   58:7  error  .throws is a syntax error  dot-notation
> 
> ✖ 1 problem (1 error, 0 warnings)

I could be fix that with something like this:
```js
t['throws'](tag.getUserAttrs);
```
But for me both look like a workaround. I mean that [`t` has `throws` method](https://github.com/substack/tape#tthrowsfn-expected-msg) and I should have an able to use it as simple as it possible.

File: lib/tag.js (hunk 4 above)
Comment (nervgh) [inline lib/tag.js]: At least twice I have looked at the code already: 
1. first time was when I was moving it from one place to another place
2. second time was when I was writing a couple of tests

I mean for me `getUserAttrs` says all what I need to know for using it.

File: test/test-tag.js (hunk 5 above)
Comment (ljharb) [inline test/test-tag.js]: That's exactly the proper solution; the tests need to work in ES3, where `.throws` is a syntax error.

Comment (nervgh) [inline test/test-tag.js]: Sorry, I did't know that the project till supports ES3. It' fixed.

File: lib/widgets.js (hunk 3 above)
Comment (nervgh) [inline lib/widgets.js]: @ljharb ping

File: lib/tag.js (hunk 4 above)
Comment (ljharb) [inline lib/tag.js]: I'm afraid it still needs all that information (that you're able to intuit) codified in explicit tests. Assume someone who has no idea what "getUserAttrs" means, wants to learn what the function does *only* by looking at its tests, including all edge cases.

**Comment by voxpelli** [comment]
Maybe we can push ahead with this @nervgh @ljharb? What needs to be fixed?

**Review by ljharb — APPROVED** [review]
